processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -127,45 +160,44 @@ protected Integer addToFacts(
       ThreadLocal<InputRow> in
   ) throws IndexSizeExceededException
   {
-    Integer rowOffset;
-    synchronized (this) {
-      rowOffset = numEntries.get();
-      if(rowOffset >= maxRowCount && !facts.containsKey(key)) {
-        throw new IndexSizeExceededException("Maximum number of rows reached");
-      }
-      final Integer prev = facts.putIfAbsent(key, rowOffset);
-      if (prev != null) {
-        rowOffset = prev;
-      } else {
-        Aggregator[] aggs = new Aggregator[metrics.length];
-        for (int i = 0; i < metrics.length; i++) {
-          final AggregatorFactory agg = metrics[i];
-          aggs[i] = agg.factorize(
-              makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
-          );
-        }
-        aggList.add(aggs);
-        numEntries.incrementAndGet();
+    final Aggregator[] aggs;
+    final Integer retval;
+    final Integer rowOffset = numEntries.get();
+    if (rowOffset >= maxRowCount && !facts.containsKey(key)) {
+      throw new IndexSizeExceededException("Maximum number of rows reached");
+    }
+    final Integer prev = facts.putIfAbsent(key, rowOffset);
```

Comment:
Put in some syncing stuff to handle the case.

Requires a nasty lock on *something* in the current implementation to make sure `aggList` and `facts` don't get out of sync.